torchbenchmark/models/soft_actor_critic/__init__.py

```diff
@@ -0,0 +1,243 @@
+            ) / self.args.num_steps
+            self.current_target_imp = lambda step: min(
+                self.args.sr_critic_target_improvement_init + self.critic_target_imp_slope * step,
+                self.args.sr_critic_target_improvement_final,
+            )
+
+    def get_module(self):
+        model = self.agent.actor
+        state = self.train_env.reset()
+        action = self.agent.sample_action(state)
+        next_state, reward, done, info = self.train_env.step(action)
+        self.buffer.push(state, action, reward, next_state, done)
+        batch = self.buffer.sample(self.args.batch_size)
+        state_batch, action_batch, reward_batch, next_state_batch, done_batch = batch
+        state_batch = state_batch.to(self.device)
+        return model, (state_batch, )
+        
+    def train(self, niter=1):
+        # Setup
+        self.target_agent.train()
+        done = True
+        for step in range(niter):
+            if done:
+                state = self.train_env.reset()
+                steps_this_ep = 0
+                done = False
+            action = self.agent.sample_action(state)
+            next_state, reward, done, info = self.train_env.step(action)
+            self.buffer.push(state, action, reward, next_state, done)
+            state = next_state
+            steps_this_ep += 1
+            if steps_this_ep >= self.args.max_episode_steps:
+                done = True
+            for _ in range(self.args.gradient_updates_per_step):
+                learn_standard(
+                    buffer=self.buffer,
+                    target_agent=self.target_agent,
+                    agent=self.agent,
+                    actor_optimizer=self.actor_optimizer,
+                    critic_optimizer=self.critic_optimizer,
+                    log_alpha=self.log_alpha,
+                    log_alpha_optimizer=self.log_alpha_optimizer,
+                    target_entropy=self.target_entropy,
+                    batch_size=self.args.batch_size,
+                    gamma=self.args.gamma,
+                    critic_clip=self.args.critic_clip,
+                    actor_clip=self.args.actor_clip,
+                    update_policy=step % self.args.actor_delay == 0,
+                    device=self.device
+                )
+
+            # move target model towards training model
+            if not self.args.self_regularized and (step % self.args.target_delay == 0):
+                soft_update(self.target_agent.critic1, self.agent.critic1, self.args.tau)
+                soft_update(self.target_agent.critic2, self.agent.critic2, self.args.tau)
+
+    def eval(self, niter=1):
+        with torch.no_grad():
+            discount= 1.0
+            episode_return_history = []
+            for episode in range(self.args.eval_episodes):
+                episode_return = 0.0
+                state = self.test_env.reset()
+                done, info = False, {}
+                for step_num in range(niter):
```

Comment:
Is there any difference between having the niter loop inside the episode loop versus outside of it?

Reply:
My guess is that the outer loop is the epoch, and the inner loop specifies how many steps we would like to run in an epoch. Therefore, I update the code so that we run `niter` epochs and each epoch we run `self.args.max_episode_steps` steps.

Original code reference:
https://github.com/jakegrigsby/deep_control/blob/2b02646c1586fd5baa23d1ee078539ccb29c16d8/deep_control/run.py#L12